### PR TITLE
Add support for RHP4 accounts

### PR DESCRIPTION
**Pull request by n8maninger**
Broken out of #490 . Requires #507

### CODE REVIEW

File: persist/sqlite/accounts.go [hunk 1]
```diff
@@ -7,12 +7,128 @@ import (
...
+		var balance types.Currency
+		err := tx.QueryRow(`SELECT id, balance FROM accounts WHERE account_id=$1`, encode(account)).Scan(&dbID, decode(&balance))
+		if err != nil {
+			return fmt.Errorf("failed to query balance: %w", err)
+		}
+
+		total := usage.RenterCost()
+		balance, underflow := balance.SubWithUnderflow(total)
+		if underflow {
+			return fmt.Errorf("insufficient balance")
```
Comment (peterjan) [inline persist/sqlite/accounts.go:38]:
```suggestion
			return errors.New("insufficient balance")
```

File: persist/sqlite/accounts.go [hunk 2]
```diff
@@ -7,12 +7,128 @@ import (
...
 	"go.sia.tech/core/types"
 	"go.sia.tech/hostd/host/accounts"
 	"go.sia.tech/hostd/host/contracts"
 	"go.uber.org/zap"
 )
 
+// RHP4AccountBalance returns the balance of the account with the given ID.
+func (s *Store) RHP4AccountBalance(account proto4.Account) (balance types.Currency, err error) {
+	err = s.transaction(func(tx *txn) error {
+		return tx.QueryRow(`SELECT balance FROM accounts WHERE account_id=$1`, encode(account)).Scan(decode(&balance))
```
Comment (peterjan) [inline persist/sqlite/accounts.go:20]: Do we want to handle `sql.ErrNoRows` and return an `ErrAccountNotFound`?

File: persist/sqlite/accounts.go [hunk 3]
```diff
@@ -7,12 +7,128 @@ import (
...
+			if err != nil && !errors.Is(err, sql.ErrNoRows) {
+				return fmt.Errorf("failed to get balance: %w", err)
+			} else if errors.Is(err, sql.ErrNoRows) {
+				createdAccounts++
+			}
+
+			balance = balance.Add(deposit.Amount)
+
+			var accountDBID int64
+			err = updateBalanceStmt.QueryRow(encode(deposit.Account), encode(balance), encode(time.Now().Add(90*24*time.Hour))).Scan(&accountDBID)
```
Comment (peterjan) [inline persist/sqlite/accounts.go:99]: `90*24*time.Hour` could be a constant

File: persist/sqlite/accounts.go [hunk 4]
```diff
@@ -7,12 +7,128 @@ import (
...
+			if err != nil {
+				return fmt.Errorf("failed to update balance: %w", err)
```
Comment (peterjan) [inline persist/sqlite/accounts.go:101]: `nit` the account ID might be useful here

File: persist/sqlite/contracts.go [hunk 5]
```diff
@@ -1121,10 +1118,32 @@ func updateV2ContractUsage(tx *txn, contractDBID int64, usage proto4.Usage) erro
 		} else if err := updateCollateralMetrics(types.ZeroCurrency, usage.RiskedCollateral, false, incrementCurrencyStat); err != nil {
 			return fmt.Errorf("failed to update collateral metrics: %w", err)
 		}
+	} else if status == contracts.V2ContractStatusSuccessful || status == contracts.V2ContractStatusRenewed {
```
Comment (peterjan) [inline persist/sqlite/contracts.go:1121]: `nit` personally prefer a switch

**Review by peterjan — COMMENTED** [review]
left some nits but good to go imo 👍

File: persist/sqlite/accounts.go [hunk 6]
```diff
@@ -212,6 +348,65 @@ func contractFunding(tx *txn, accountID int64) (fund []fundAmount, err error) {
 	return
 }
 
+// updateV2ContractFunding distributes account usage to the contracts that funded
+// the account.
+func updateV2ContractFunding(tx *txn, accountID int64, usage proto4.Usage) error {
```
Comment (ChrisSchinnerl) [inline persist/sqlite/accounts.go:353]: This seems a bit complicated for a solution that results in inaccurate information. It might be simpler and more accurate to always credit the last contract that was used for funding an account.
Or to not associate usage with contracts and instead with accounts.

Comment (ChrisSchinnerl) [reply persist/sqlite/accounts.go:353]: Also the name is not consistent with `updateContractUsage` which seems to be the same thing but for v1?

File: persist/sqlite/accounts.go [hunk 7]
```diff
@@ -265,7 +460,7 @@ func updateContractUsage(tx *txn, accountID int64, usage accounts.Usage, log *za
 			return fmt.Errorf("failed to decrement account funding: %w", err)
 		}
 
-		if contract.Status == contracts.ContractStatusActive || contract.Status == contracts.ContractStatusPending {
+		if contract.Status == contracts.ContractStatusActive {
```
Comment (ChrisSchinnerl) [inline persist/sqlite/accounts.go:463]: What triggered this change? Isn't this pretty much guaranteed to miss metrics since we start using accounts before the contract appears on chain?

File: persist/sqlite/accounts_test.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+			t.Fatalf("expected storage %v, got %v", earned.Storage, m.Revenue.Earned.Storage)
+		case m.Revenue.Earned.RPC != earned.RPC:
+			t.Fatalf("expected RPC %v, got %v", earned.RPC, m.Revenue.Earned.RPC)
+		}
+	}
+
+	sk := types.GeneratePrivateKey()
+	account := proto4.Account(sk.PublicKey())
+
+	// deposit funds
```
Comment (ChrisSchinnerl) [inline persist/sqlite/accounts_test.go:80]: I think `RHP4AccountBalance` is not called at all and we should probably also test the edge cases of withdrawing from an account and fetching the balance of an account before it even exists.

File: persist/sqlite/accounts.go (hunk 4 above)
Comment (n8maninger) [reply persist/sqlite/accounts.go:101]: account ID doesn't matter if the query fails. It's unlikely to be related to the account ID.

File: persist/sqlite/accounts.go (hunk 6 above)
Comment (n8maninger) [reply persist/sqlite/accounts.go:353]: Associating usage with an account is unacceptable UX.

File: persist/sqlite/accounts.go (hunk 7 above)
Comment (n8maninger) [reply persist/sqlite/accounts.go:463]: The contract metrics are added to the global metrics when the status changes from pending -> active, which is consistent with the contract revision PR that just merged. This was a necessary change when finalization was a thing. Now that it's not we could revert it, but it does simplify the edge-cases around pending->reverted->confirmed.

File: persist/sqlite/accounts.go (hunk 6 above)
Comment (ChrisSchinnerl) [reply persist/sqlite/accounts.go]: The UX I imagine is 
1. click on contract
2. contract shows you contract spending and a list of accounts and amounts that it was used to fund with a number indicating the amount. Most of the time that will be a single account.
3. you click the account to get detailed spending metrics for it and a list of contracts that were used to fund it for you to seamlessly navigate between contracts and accounts

If you spread it out across contracts you end up with some awkward UX depending on whether you also include expired contracts or not. Either you increase the spending of contracts which are expired which isn't great or you potentially end up seemingly spending more money from a contract than it contained to begin with.

Comment (n8maninger) [reply persist/sqlite/accounts.go]: That UX is significantly more complex than associating it with the contract  in the first place, which is a more accurate representation and what users are expecting.

**Review by ChrisSchinnerl — APPROVED** [review]
